### PR TITLE
search contexts: resolve namespaced contexts

### PR DESCRIPTION
**Pull request by novoselrok**
Added support for resolving namespaced contexts: `@user/context`, `@org/context`, etc.

To avoid overloading the DB when constructing search context specs (e.g. when listing search contexts), I also cache the namespace names (username and org name) in the struct.

### CODE REVIEW

File: cmd/frontend/internal/search/searchcontexts/search_contexts.go [hunk 1]
```diff
@@ -20,19 +20,34 @@ const (
 	searchContextSpecPrefix = "@"
 )
 
+var namespacedSearchContextSpecRegexp = regexp.MustCompile(searchContextSpecPrefix + `(.*?)\/(.*)`)
```
Comment (camdencheek) [inline cmd/frontend/internal/search/searchcontexts/search_contexts.go:23]: Regexes in the backend should use `lazyregexp`

File: cmd/frontend/internal/search/searchcontexts/search_contexts.go [hunk 2]
```diff
@@ -20,19 +20,34 @@ const (
...
+
 func ResolveSearchContextSpec(ctx context.Context, db dbutil.DB, searchContextSpec string) (*types.SearchContext, error) {
 	if IsGlobalSearchContextSpec(searchContextSpec) {
 		return GetGlobalSearchContext(), nil
+	} else if submatches := namespacedSearchContextSpecRegexp.FindStringSubmatch(searchContextSpec); len(submatches) == 3 {
```
Comment (camdencheek) [inline cmd/frontend/internal/search/searchcontexts/search_contexts.go:28]: nit: general convention for checking if the pattern matches is to check `submatches != nil`

File: internal/database/search_contexts.go [hunk 3]
```diff
@@ -233,8 +234,11 @@ func (s *SearchContextsStore) createSearchContext(ctx context.Context, searchCon
 	if err != nil {
 		return nil, err
 	}
-	defer rows.Close()
-	return scanSingleSearchContext(rows)
+	return s.GetSearchContext(ctx, GetSearchContextOptions{
```
Comment (camdencheek) [inline internal/database/search_contexts.go:237]: Why was this changed to two DB roundtrips rather than using `RETURNING`? I guess it doesn't matter too much since search contexts are created rarely, but still curious

File: cmd/frontend/internal/search/searchcontexts/search_contexts.go [hunk 4]
```diff
@@ -20,19 +20,34 @@ const (
 	searchContextSpecPrefix = "@"
 )
 
+var namespacedSearchContextSpecRegexp = regexp.MustCompile(searchContextSpecPrefix + `(.*?)\/(.*)`)
+
 func ResolveSearchContextSpec(ctx context.Context, db dbutil.DB, searchContextSpec string) (*types.SearchContext, error) {
```
Comment (camdencheek) [inline cmd/frontend/internal/search/searchcontexts/search_contexts.go:25]: Should we consider making a type for `searchContextSpec`? It feels a little weird to be parsing it this late now that there are so many supported formats. Also might be useful for the `GetSearchContextSpec` function

**Review by camdencheek — APPROVED** [review]
Looks nice. Just a few small style comments

File: internal/database/search_contexts.go (hunk 3 above)
Comment (novoselrok) [reply internal/database/search_contexts.go:237]: Because when retrieving the search contexts, we now join users and orgs tables so we can cache the username and org name. The `RETURNING` clause unfortunately does not allow that. And as you've mentioned, this is not exactly performance critical, so it should be fine.

File: cmd/frontend/internal/search/searchcontexts/search_contexts.go (hunk 4 above)
Comment (novoselrok) [reply cmd/frontend/internal/search/searchcontexts/search_contexts.go:25]: Yep, that might be a good idea. I'll add a note to the search context project, so we don't forget about it.
